src/effects/effectchainslot.cpp

```diff
@@ -113,13 +136,305 @@ EffectChainSlot::~EffectChainSlot() {
+        m_pEffectsManager->registeredInputChannels(),
+        m_pEffectsManager->registeredOutputChannels());
+    EffectsRequest* pRequest = new EffectsRequest();
+    pRequest->type = EffectsRequest::ADD_CHAIN_TO_RACK;
+    pRequest->pTargetRack = pRack;
+    pRequest->AddChainToRack.pChain = m_pEngineEffectChain;
+    pRequest->AddChainToRack.iIndex = iIndex;
+    m_pEffectsManager->writeRequest(pRequest);
+
+    // Add all effects.
+    for (int i = 0; i < m_effects.size(); ++i) {
+        // Add the effect to the engine.
+        EffectPointer pEffect = m_effects[i];
+        if (pEffect) {
+            pEffect->addToEngine(m_pEngineEffectChain, i, m_enabledInputChannels);
+        }
+    }
+}
+
+void EffectChainSlot::removeFromEngine(EngineEffectRack* pRack, int iIndex) {
+    // Order doesn't matter when removing.
+    for (int i = 0; i < m_effects.size(); ++i) {
+        EffectPointer pEffect = m_effects[i];
+        if (pEffect) {
+            pEffect->removeFromEngine(m_pEngineEffectChain, i);
+        }
+    }
+
+    EffectsRequest* pRequest = new EffectsRequest();
+    pRequest->type = EffectsRequest::REMOVE_CHAIN_FROM_RACK;
+    pRequest->pTargetRack = pRack;
+    pRequest->RemoveChainFromRack.pChain = m_pEngineEffectChain;
+    pRequest->RemoveChainFromRack.iIndex = iIndex;
+    m_pEffectsManager->writeRequest(pRequest);
+
+    m_pEngineEffectChain = nullptr;
+}
+
+void EffectChainSlot::updateEngineState() {
+    // Update chain parameters in the engine.
+    sendParameterUpdate();
+    for (int i = 0; i < m_effects.size(); ++i) {
+        EffectPointer pEffect = m_effects[i];
+        if (pEffect) {
+            // Update effect parameters in the engine.
+            pEffect->updateEngineState();
+        }
+    }
+}
+
+const QString& EffectChainSlot::name() const {
+    return m_name;
+}
+
+void EffectChainSlot::setName(const QString& name) {
+    m_name = name;
+    // emit(nameChanged(name));
+    slotChainNameChanged(name);
+}
+
+QString EffectChainSlot::description() const {
+    return m_description;
+}
+
+void EffectChainSlot::setDescription(const QString& description) {
+    m_description = description;
+    // emit(descriptionChanged(description));
+}
+
+bool EffectChainSlot::enabled() const {
+    return m_bEnabled;
+}
+
+void EffectChainSlot::setEnabled(bool enabled) {
+    m_bEnabled = enabled;
+    sendParameterUpdate();
+    // emit(enabledChanged(enabled));
+    slotChainEnabledChanged(enabled);
+}
+
+void EffectChainSlot::enableForInputChannel(const ChannelHandleAndGroup& handle_group) {
+    // TODO(Be): remove m_enabledChannels from this class and move this logic
+    // to EffectChainSlot
+    bool bWasAlreadyEnabled = m_enabledInputChannels.contains(handle_group);
+    if (!bWasAlreadyEnabled) {
+        m_enabledInputChannels.insert(handle_group);
+    }
+
+    // The allocation of EffectStates below may be expensive, so avoid it if
+    // not needed.
+    if (bWasAlreadyEnabled) {
+        return;
+    }
```

Comment:
Can this check be removed at this point? IIRC it was only necessary before because of the convoluted initialization process.